2-0-data-structures-and-algorithms/src/test/java/com/bobocode/bst/BinarySearchTreeTest.java

```diff
@@ -1,82 +1,249 @@
+    void shouldNotInsertNonUniqueElements() {
+        fillTestTree(someElements);
+
+        assertThat(tree.insert(10)).isFalse();
+        assertThat(tree.insert(11)).isFalse();
+        assertThat(tree.insert(12)).isFalse();
+    }
+
+    @Test
+    @Order(9)
+    void shouldNotIncreaseSizeWhenInsertNotUnique() {
+        fillTestTree(someElements);
+
+        assertThat(getInnerSize()).isEqualTo(someElements.length);
+
+        tree.insert(10);
+        tree.insert(11);
+        tree.insert(12);
+
+        assertThat(getInnerSize()).isEqualTo(someElements.length);
+    }
+
+    @Order(10)
     @ParameterizedTest
     @MethodSource("depthArguments")
-    @Order(4)
     void depth(Integer[] elements, int depth) {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of(elements);
-        assertThat(bst.depth()).isEqualTo(depth);
+        fillTestTree(elements);
+        assertThat(tree.depth()).isEqualTo(depth);
     }
 
     @Test
-    @Order(5)
+    @Order(11)
     void inorderTraversal() {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of(someElements);
+        fillTestTree(someElements);
         Integer[] sortedElements = Arrays.copyOf(someElements, someElements.length);
         Arrays.sort(sortedElements);
 
-        List<Integer> traversedElements = new ArrayList<>(bst.size());
-        bst.inOrderTraversal(traversedElements::add);
+        List<Integer> traversedElements = new ArrayList<>(getInnerSize());
+        tree.inOrderTraversal(traversedElements::add);
 
         assertThat(traversedElements).isEqualTo(List.of(sortedElements));
     }
 
+    @Test
+    @Order(12)
+    void shouldInsertToRootIfEmpty() {
+        tree = new RecursiveBinarySearchTree<>();
+        assertThat(getRootObject()).isNull();
+        tree.insert(10);
+        assertThat(getRootObject()).isNotNull();
+    }
+
+    @Test
+    @Order(13)
+    void shouldInsertLeftIfLessThanRoot() {
+        tree = new RecursiveBinarySearchTree<>();
+
+        tree.insert(10); //root
+        tree.insert(5); //left
+
+        assertThat(getLeftNode(getRootObject())).isNotNull();
+    }
+
+    @Test
+    @Order(14)
+    void shouldInsertRightIfGreaterThanRoot() {
+        tree = new RecursiveBinarySearchTree<>();
+
+        tree.insert(10); //root
+        tree.insert(15); //right
+
+        assertThat(getRightNode(getRootObject())).isNotNull();
+    }
+
+    @Test
+    @Order(15)
+    void depthShouldGrowWhenInsert() {
+        tree = new RecursiveBinarySearchTree<>();
+
+        tree.insert(13);
+        tree.insert(11);
+        tree.insert(12);
+        tree.insert(15);
+        tree.insert(-15);
+
+        assertThat(tree.depth()).isGreaterThan(0);
+    }
+
+    @Test
+    @Order(16)
+    void depthShouldBeZeroIfRootNull () {
+        tree = new RecursiveBinarySearchTree<>();
+        assertThat(tree.depth()).isEqualTo(0);
+    }
+
+    @Test
+    @Order(17)
+    void shouldReturnTrueIfContains() {
+        fillTestTree(someElements);
```

Comment:
@erneman1 I suggest that `fillTestTree()` **use varargs parameters**. Then the test will look more obvious since you will be able to see which values have been added.

```java
    fillTestTree(10, 9, 11);

    assertThat(tree.contains(10)).isTrue();
    assertThat(tree.contains(9)).isTrue();
    assertThat(tree.contains(11)).isTrue();
```